KEP-3063: DRA: 1.30 update (#4181)
kubernetes/enhancements: keps/sig-node/3063-dynamic-resource-allocation/README.md

    - [Reserve](#reserve)

>     - [Unreserve](#unreserve)

Reviewer: I believe you are working on some improvements in scheduler in this cycle. Can you please summarize in the KEP?

Author: The event handling changes already went into 1.28. I'll check how to document that here.

We were discussing moving the blocking API calls out of the main scheduling cycle, but I am with @Huang-Wei: let's not do that and instead just document the effect. Will add that here.

  name: acme-gpu-init
```

#### Integration with cluster autoscaling

> As a cloud provider, I want to support GPUs as part of a hosted Kubernetes

Reviewer 2: I don't like the implications of this user story - I think it implies the DRA will only ever work in hosted environments such as GKE and will not be available in self-hosted clusters running it cloud (you even explicitly mention hosted Kubernetes environment). And it also implies every provider will likely need to their own version of integration, resulting in inconsistent experience for anyone using multi-cloud.

Can we explore what would happen if we changed this to "as a hardware vendor, I want to make my hardware available for applications that run in autoscaled cloud environments". This obviously makes it much harder to add a plugin in the first place, but it's also the most realistic way to guarantee support across cloud environments.

environment, including cluster autoscaling. I ensure that the kernel is
configured as required by the hardware and that the container runtime supports

> CDI. I review the Go code provided by the vendor for simulating cluster scaling

Reviewer 2: AFAIK vast majority of users use the default Cluster Autoscaler image. Certainly not every cloud provider supports their own image - I don't have any hard data here, but I believe most don't.

- *Resource kubelet plugin*: a component which cooperates with kubelet to prepare
  the usage of the resource on a node.

> An [utility library](https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/dynamic-resource-allocation) for resource drivers was developed.

Author: Applied, will be in next update.

	// than one consumer at a time.
	// +optional
	Shareable bool

> 	<<[UNRESOLVED pohly]>>

Author: I'll add it to alternatives. Basically no-one has come up with use cases for more complex sharing scenarios, so we are keeping it simple (shared/not shared).

needs to be put back into the work queue. It then gets retried whenever a
ResourceClaim gets added or modified.

The following extension points are implemented in the new claim plugin. Some of
them invoke API calls to create or update objects. This is done to simplify
error handling: a failure during such a call puts the pod into the backoff

> queue where it will be retried after a timeout. The downside is that the

Reviewer 3: This is troubling. I think it needs more thought. I think not disrupting scheduling of other pods needs to be a graduation criteria.

Author: It's already almost solved: https://github.com/kubernetes/kubernetes/pull/121876

A discussion around PostFilter is on-going on Slack.

Normally, architectural changes would be first described in a KEP and then get implemented, but changes to existing code are easier to explore, show and comment on via a PR. Let's update this text once the code discussions are done.

/hold

> I think not disrupting scheduling of other pods needs to be a graduation criteria.

Agreed.

Author: Added and text updated.

We decided to not invest undue effort (and introduce complexity) for addressing the PostFilter deallocate edge case.

Author: /hold cancel

needs to be put back into the work queue. It then gets retried whenever a
ResourceClaim gets added or modified.

The following extension points are implemented in the new claim plugin. Except
for some unlikely edge cases (see below) there are no API calls during the main
scheduling cycle. Instead, the plugin collects information and updates the
cluster in the separate goroutine which invokes PreBind.

#### EventsToRegister

This registers all cluster events that might make an unschedulable pod
schedulable, like creating a claim that the pod needs or finishing the
allocation of a claim.

> [Queuing hints](https://github.com/kubernetes/enhancements/issues/4247) are

Reviewer 4: +1. For this KEP update, we will have to state the status of DRA w/o usage of queueing hints.

> [Queuing hints](https://github.com/kubernetes/enhancements/issues/4247) are

Author: It's not crucial. Performance is different, but DRA is usable also without that optimization. I know that you disagree, but all the use cases in https://docs.google.com/document/d/1XNkTobkyz-MyXhidhTp5RfbMsM-uRCWDoflUMqNcYTk/edit#heading=h.ljj9kaa144nr confirmed that scheduling performance is not a priority because of long-running pods.

I'll update the text to explain this.

> [Queuing hints](https://github.com/kubernetes/enhancements/issues/4247) are

Author: I added something.

supported. These are callbacks that can limit the effect of a cluster event to
specific pods. For example, allocating a claim only makes those pods
scheduleable which reference the claim. There is no need to try scheduling a pod
which waits for some other claim. Hints are also used to trigger the next
scheduling cycle for a pod immediately when some expected and require event
like "drivers have provided information" occurs, instead of forcing the pod to
go through the backoff queue and the usually 5 second long delay associated
with that.

> #### PreEnqueue

Reviewer 4: FWIW, PreEnqueue is likely to reach GA in 1.30.

encounters a pod that uses a resource claim, the autoscaler needs assistance by
the resource driver for that claim to make the right decisions. Without that
assistance, the autoscaler might scale up the wrong node group (resource is
provided by nodes in another group) or not scale up (pod is pending because of
a claim that cannot be allocated, but looks like it should be scheduleable
to the autoscaler).

With the following changes, vendors can provide Go code in a package that can

> be built into a custom autoscaler binary to support correct scale up

Author: See my new sections (added just now) about this being required only for node-local resources. If a cloud provider rolls out a new machine type with special hardware that needs a DRA driver, then I expect that they are also prepared to provide support for that hardware at the software level.

Can someone from SIG Autoscaling comment on this (@mwielgus, @x13n)? You know best how downstream users are consuming the "official" cluster autoscaler. Is it common that they run some modified version?

> be built into a custom autoscaler binary to support correct scale up

Reviewer 5: I wrote a similar comment at the first mention of this, but am jumping on here instead.

This is kind of horrible.  We already have a mess with multiple implementations of cluster-autoscaling, some open-source and some not (which I care less about, but not _zero_).   This model seems problematic for multiple reasons.  First, we have to define an API that works across some majority of CA implementations (hard!).  Second, we have to support this as a permanent API, forever.  Third, we're asking cluster admins to build their own binaries - if they have problems, nobody in upstream will want to touch them until they can repro without the plugins.

That makes it sound like I am against some form of loadable modules - I am not.  I just know how complicated they are and think we should go there VERY cautiously.

More than that, I don't think we can support so many different extension mechanisms for the same subsystem.  As a producer of a plugin I have to deal with:
* CRDs
* PodSchedulingContext
* gRPC to per-node driver
* Now compiled-in "modules" ?

This just doesn't smell right to me.

> be built into a custom autoscaler binary to support correct scale up

Reviewer 2: > You know best how downstream users are consuming the "official" cluster autoscaler. Is it common that they run some modified version?

It's hard to tell for sure how people run it across clouds, but AFAIK vast majority of users just use the official image. Some managed solutions (ex. GKE, AKS) either host cluster-autoscaler or provide tooling for managing it, I guess at least some of them build their own images. Outside of that I think it's very rare - CA is built to be easily extensible and there are some parts (expander) which are both very easy to customize and come with clear incentive for customization (tailoring decision logic to specific use-case) and yet I have yet to see anyone mention running a CA with a custom expander.

simulations for clusters that use their hardware. Extensions for invoking such
vendor code through some RPC mechanism, as WASM plugin, or some generic

> code which just needs to be parameterized for specific hardware could be added

Reviewer 5: +1 - I called this KEP "terrifying" at some point in the past.  I don't want to be paralyzed by fear, but we really don't know what we don't know yet.

> code which just needs to be parameterized for specific hardware could be added

Author: Obviously the proposal for autoscaling needs more discussion and therefore we are not meeting the beta criteria. Let's keep this feature in alpha for 1.29 and use that time to continue the discussion.

But this PR contains several useful enhancements, so I propose that we merge it without the "graduation to beta" part. I removed that from kep.yaml. I also added an `UNRESOLVED` paragraph to the autoscaler section to make it clear that even if this gets merged, there is no consensus that this is how it should be. Is that okay?

Here's what I added:

> The entire autoscaler section is tentative. Key opens:
> 
> - Are DRA driver authors able and willing to provide implementations of
>   the simulation interface?
> - Is the simulation interface generic enough to work across a variety
>   of autoscaler forks and/or implementations? What about Karpenter?
> - Is the suggested deployment approach (rebuild binary) workable?
> - Can we really not do something else, ideally RPC-based?

I can also leave the entire section unchanged in this PR, if that's preferred.

//   points that the plugin is being used as part of the cluster autoscaler.
// . In PreFilter and Filter use the cluster snapshot to make decisions
//   instead of the normal "live" cluster state.
// - In SimulateBindPod update the snapshot in the cycle state.
type ClusterAutoScalerPlugin interface {
	Plugin
	// StartSimulation is called when the cluster autoscaler begins
	// a simulation.
	StartSimulation(ctx context.Context, state *CycleState) *Status
	// SimulateBindPod is called when the cluster autoscaler decided to schedule
	// a pod onto a certain node.
	SimulateBindPod(ctx context.Context, state *CycleState, pod *v1.Pod, nodeInfo *NodeInfo) *Status
	// SimulateEvictPod is called when the cluster autoscaler simulates removal
	// of a node. All claims used only by this pod should be considered deallocated,
	// to enable starting the same pod elsewhere.
	SimulateEvictPod(ctx context.Context, state *CycleState, pod *v1.Pod, nodeName string) *Status
    // NodeIsReady checks whether some real node has been initialized completely.
    // Even if it is "ready" as far Kubernetes is concerned, some DaemonSet pod
    // might still be missing or not done with its startup yet.

>     NodeIsReady(ctx context.Context, node *v1.Node) (bool, error)

Reviewer: @MaciekPytel shouldn't this by done by cluster-autoscaler, outside of any extension?

Author: How would CA know which daemon set pods are needed and whether they have finished their startup work so that claims can be allocated for the node? It can't.

There are some hacks specifically for GPUs with extended resources in the current code base for this, but that won't work for DRA drivers.

Reviewer: IIUC, all that CA needs to know is whether the Node will be able to eventually accommodate the Pod. So here it needs to know which Daemonsets Pods will be required if they aren't already there. This sounds independent of the driver.
In fact, knowing that a Node is not ready (right now) could potentially cause unnecessary scale ups.

Reviewer 2: This is solved for GPU by knowing that a GPU node will always have a pre-defined (provider-specific) label and a gpu capacity. Autoscaler knows that a node that has the label, but no GPU capacity is still booting up. 
Another popular solution for this problem is using a 'startup-taint' - a node starts with a taint and the daemonset removes the taint once it's done with whatever intialization it needs to do. Startup taints are passed to CA via a flag or matched via regexp.

Finally, there is an [existing extension point](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/processors/customresources/custom_resources_processor.go) within autoscaler that can be used to implement this based on any other criteria (GPU and taints use this extension point internally).

Reviewer 6: With DRA, a simple "capacity" is not enough for GPUs. The API for requesting GPUs is much more complex:
https://github.com/NVIDIA/k8s-dra-driver/blob/main/api/nvidia.com/resource/gpu/v1alpha1/gpuclaim.go#L26

Reviewer 2: It doesn't need to be simple capacity for the model to work. The way it works today is actually pretty much the same as what is proposed here. The extension point I mentioned requires implementing `FilterOutNodesWithUnreadyResources` function which boils down to being able to tell if a node is fully initialized or not. And the taint-based solution can work with any initialization, it only requires that the node is started with a taint and the process doing the initialization removes it at the end.

We can add other mechanisms if needed. The only real requirement is that there are some clear, predefined criteria that let autoscaler tell if a node is ready or not based just on looking on the node object.

When [Cluster
Autoscaler](https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler#cluster-autoscaler)
encounters a pod that uses a resource claim for node-local resources, it needs
to understand the parameters for the claim and available capacity in order
to simulate the effect of allocating claims as part of scheduling and of
creating or removing nodes.

This is not possible with opaque parameters as described in this KEP. If a DRA
driver developer wants to support Cluster Autoscaler, they have to use numeric
parameters. Numeric parameters are an extension of this KEP that is defined in
[KEP #4381](https://github.com/kubernetes/enhancements/issues/4381).

> Numeric parameters are not necessary for network-attached resources because

Reviewer 3: I think it's an assumption that we would only ever autoscale nodes. But probably a safe one for the time being.

Author: I was thinking specifically of the current Cluster Autoscaler. I agree that some other (current or future) implementation might do more, but we have to limit the scope somewhere or we'll get paralyzed by everything that might come...

prioritize which resource to deallocate first. Future extensions of this KEP
might attempt to improve this.

> This is currently using blocking API calls. They are unlikely because this

Reviewer 4: ```suggestion
This is currently using blocking API calls. It's quite rare because this
```

### Cluster Autoscaler

> -<<[UNRESOLVED pohly]>>

Reviewer: FYI, I want to clarify that, given that this entire section is marked as UNRESOLVED, and sig-autoscaling seems mostly opposed to the idea of compiled-in plugins, I won't be accepting the code changes for kube-scheduler suggested in this section.

You can still iterate on the design during the release.

Author: As discussed after KubeCon NA 2023 and in Batch WG, the intended solution for cluster autoscaling is to make parameters less opaque via "numeric parameters". This is, at least from a technical perspective, an extension of the core DRA. Therefore I have created a [separate KEP #4381](https://github.com/kubernetes/enhancements/issues/4381) for it and in this PR pushed an update that makes it clear that cluster autoscaling will not work as expected when using opaque parameters.

If we agree on this approach, then this PR might be ready for merging.

https://github.com/kubernetes/enhancements/pull/4181#issuecomment-1750755469:

> I would like to see an exploration of how those two concepts can be detached.

I've taken this into account during the design of numeric parameters. Let's head over to the [initial PR for numeric parameters](https://github.com/kubernetes/enhancements/pull/4384) and continue there...

influences further checks for other pending pods. During scale down, eviction
is simulated by
[SimulateNodeRemoval](https://github.com/kubernetes/autoscaler/blob/2f7c61e13bd1cbfc0ba4085fb84bd692a1e9ac6e/cluster-autoscaler/simulator/cluster.go#L149)
which [pretends that pods running on a node that is to be removed are not
running](https://github.com/kubernetes/autoscaler/blob/2f7c61e13bd1cbfc0ba4085fb84bd692a1e9ac6e/cluster-autoscaler/simulator/cluster.go#L231-L237).

The DRA scheduler plugin gets integrated into this snapshotting and simulated
pod scheduling through a new scheduler framework interface:

```
// ClusterAutoScalerPlugin is an interface that is used only by the cluster autoscaler.
// It enables plugins to store state across different scheduling cycles.
//
// The usual call sequence of a plugin when used in the scheduler is:
// - at program startup:
//   - instantiate plugin
//   - EventsToRegister
// - for each new pod:
//   - PreEnqueue
// - for each pod that is ready to be scheduled, one pod at a time:
//   - PreFilter, Filter, etc.
//

> // Cluster autoscaler works a bit differently. It identifies all pending pods,

Reviewer 5: What about Karpenter?  It's got a lot of attention lately, and I think we need to consider how it would engage with this feature.  There are others, but I don't hear much about them - we may need to do outreach.

Reviewer 2: @ellistarn

AFAIK Karpenter doesn't import scheduler plugins the way CAS does. Instead it re-implements the underlying logic in its own framework (keep me honest here @ellistarn), which would require yet another re-implementation of the plugin logic? Maybe an RPC call would be more feasible in Karpenter than it is in CAS?

Reviewer 7: Yeah we don't currently support scheduler plugins. Let me read up a bit and form an opinion on this one.

Reviewer 7: I'm not actually that concerned about the scheduler implementation. It's a much longer discussion, but due to the higher dimensionality of karpenter than CAS, we can't use the schedulers "search" approach. 

I am a bit sad that this KEP overhauls extended resources, but fails to address a critical issue that NodeAutoscalers face with node capacity discovery. CAS faces this with scale to zero, where a users must manually tag their asgs (or an equivalent vendor specific analogue) with the extended resources that will be attached to the node. 

Karpenter faces a similar problem, where we can't know that a node will have an extended resource until a node launches. The AWS cloud provider currently hardcodes knowledge of a small set of extended resources (Nvidia, etc) that are most relevant to our customers, but similar to CAS discovery, we will need to build a discovery mechanism to support 3rd party plugins.

Reviewer 5: As prevalent as autoscaling is, I think it is safe to say that this feature is open to new ideas, up to and including "throw the whole thing away and start over", if needed.

permanent deadlock that somehow would have to be detected and resolved to make
progress.

Unreserve is called in two scenarios:
- In the main goroutine when scheduling a pod has failed: in that case the plugin's
  Reserve call hasn't actually changed the claim status yet, so there is nothing
  that needs to be rolled back.
- After binding has failed: this runs in a goroutine, so reverting the
  `claim.status.reservedFor` with a blocking call is acceptable.

### Cluster Autoscaler

When [Cluster
Autoscaler](https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler#cluster-autoscaler)
encounters a pod that uses a resource claim for node-local resources, it needs
to understand the parameters for the claim and available capacity in order
to simulate the effect of allocating claims as part of scheduling and of
creating or removing nodes.

This is not possible with opaque parameters as described in this KEP. If a DRA
driver developer wants to support Cluster Autoscaler, they have to use numeric

> parameters. Numeric parameters are an extension of this KEP that is defined in

Reviewer: I don't think we should phrase this case as only being an enabler for cluster autoscaler. I see that KEP as an opportunity to simplify this KEP before it goes to beta.

Author: I can add that.

Author: It didn't fit into this "Autoscaler" section, so I added a new section for this thought under "Alternatives".

###### What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?

> For kube-controller-manager, metrics similar to the generic ephemeral volume

Reviewer: Are there any metrics we can add to scheduler to better understand some internal steps?

Even just a simple counter for creating the new API objects (like PodScheduling) could be useful. Or a gauge for the number of pods that are waiting for updates in the objects before being scheduled.

Author: > Even just a simple counter for creating the new API objects (like PodScheduling) could be useful.

I agree, this seems useful. I'll add that here and will implement for 1.29.

> Or a gauge for the number of pods that are waiting for updates in the objects before being scheduled.

I think this is covered by ["unschedulable_pods"](https://github.com/kubernetes/kubernetes/blob/6f5fa2eb2f4dc731243b00f7e781e95589b5621f/pkg/scheduler/metrics/metrics.go#L200-L206). Let me describe that here.

Author: > I agree, this seems useful. I'll add that here and will implement for 1.29.

I just remembered why I didn't already add those: the same information about "PodSchedulingContexts created/updated" is already available through the apiserver metrics. IMHO there is no need to also record the same thing on the client side. The only difference would be that a "failure" count could also include requests that don't even reach the apiserver. Not sure whether that's worth it?

For now I'll add documentation about what exists already.

###### What are other known failure modes?

- DRA driver does not or cannot allocate a resource claim.

  - Detection: The primary mechanism is through vendors-provided monitoring for
    their driver. That monitor needs to include health of the driver, availability
    of the underlying resource, etc. The common helper code for DRA drivers
    posts events for a ResourceClaim when an allocation attempt fails.

    When pods fail to get scheduled, kube-scheduler reports that through events
    and pod status. For DRA, that includes "waiting for resource driver to

>     provide information" (node not selected yet) and "waiting for resource

Reviewer 3: Do these events/status updates include information about the specific claim/driver that is blocking progress?

Author: When waiting for a claim, the claim is mentioned. This allows the user to drill down and check the claim.

When waiting for PodSchedulingContext information, that object is not mentioned explicitly. It doesn't need to be named because the name is the same as for the pod.

In both cases it is assumed that users understand the concepts enough to know what "claim" and "pod scheduling context" are.

## Alternatives

### Semantic Parameters instead of PodSchedulingContext

> When a DRA driver uses semantic parameters, there is no DRA driver controller

Reviewer 3: I believe in the latest cut this has shifted a little? In the semantic parameter version, there is still a driver controller but its responsibility is to evaluate the claim parameters and produce the driver-neutral semantic request. Since we are skipping CEL for this (for now at least...). But the following text on "we might be able to remove PodSchedulingContext" looks right.

Author: I change the text into "there is no need for a DRA driver controller
which allocates the claim and no need for communication between scheduler and such a controller"

That leaves it open whether a controller is still needed for other purposes (will be defined in the "semantic parameters KEP" and may depend on whether users are allowed to create in-tree parameter objects directly) and just focuses on the aspect that is relevant here.

}
```

### Complex sharing of ResourceClaim

> At the moment, the allocation result marks as a claim as either "shareable" by

Reviewer 3: There's something not quite right in that model. I wonder if the underlying semantic models can express this?

Also, there are different kinds of "shareable" - for example, I know of use cases where you want to mount the GPU devices in a monitoring container - or perhaps to access it via some config tools.

No change needed, just raising some things to think about post 1.30.
